source/codegen/templates/service_helpers.mako

```diff
@@ -271,6 +271,227 @@ ${populate_response(function_data=function_data, parameters=parameters)}\
+            [&](auto x) {
+                return x;
+            });
+        packedData.PackFrom(function_data->data);
+    }
+% else:
+    if (status >= 0) {
+        std::transform(
+            array,
+            array + ${size},
+            function_data->data.mutable_value()->begin(),
+            [&](auto x) {
+                return x;
+            });
+        packedData.PackFrom(function_data->data);
+    }
+% endif
+</%def>
+
+<%def name="streaming_handle_out_direction_scaler(c_api_name, arg_string, streaming_type)">\
+${streaming_type} value = 0;
+    auto status = library->${c_api_name}(${arg_string});
+    function_data->data.set_value(value);
+    if (status >= 0) {
+        packedData.PackFrom(function_data->data);
+    }
+</%def>
+
+<%def name="streaming_handle_in_direction(c_api_name, arg_string, data_type, grpc_streaming_type, streaming_type, streaming_param)">
+<%
+  is_array = common_helpers.is_array(streaming_type)
+%>\
+    ${grpc_streaming_type} ${grpc_streaming_type.lower()}_message;
+    packedData.UnpackTo(&${grpc_streaming_type.lower()}_message);
+% if is_array:
+    ${streaming_handle_in_direction_array(data_type, grpc_streaming_type, streaming_param)}\
+% else:
+    ${streaming_handle_in_direction_scaler(grpc_streaming_type, streaming_type, streaming_param)}\
+% endif
+
+    auto status = library->${c_api_name}(${arg_string});
+</%def>
+
+<%def name="streaming_handle_in_direction_scaler(grpc_streaming_type, streaming_type, streaming_param)">\
+<%
+   is_coerced = service_helpers.is_scalar_input_that_needs_coercion(streaming_param)
+%>\
+auto value = ${grpc_streaming_type.lower()}_message.value();
+% if is_coerced:
+    if (value < std::numeric_limits<${streaming_type}>::min() || value > std::numeric_limits<${streaming_type}>::max()) {
+      std::string message("value " + std::to_string(value) + " doesn't fit in datatype ${streaming_type}");
+      throw nidevice_grpc::ValueOutOfRangeException(message);
+    }
+% endif
+</%def>
+
+<%def name="streaming_handle_in_direction_array(data_type, grpc_streaming_type, streaming_param)">\
+<%
+   is_coerced = service_helpers.is_input_array_that_needs_coercion(streaming_param)
+%>\
+% if is_coerced:
+auto data_array = ${grpc_streaming_type.lower()}_message.value();
+    auto array = std::vector<${data_type}>();
+    auto size = data_array.size();
+    array.reserve(size);
+    std::transform(
+        data_array.begin(),
+        data_array.end(),
+        std::back_inserter(array),
+        [](auto x) {
+          if (x < std::numeric_limits<${data_type}>::min() || x > std::numeric_limits<${data_type}>::max()) {
+            std::string message("value " + std::to_string(x) + " doesn't fit in datatype ${data_type}");
+            throw nidevice_grpc::ValueOutOfRangeException(message);
+          }
+          return static_cast<${data_type}>(x);
+        });
+% else:
+    auto data_array = ${grpc_streaming_type.lower()}_message.value();
```

Comment:
Consider if any of your `auto`s should be `const auto x&` or some other qualifiers to avoid copy on assignment and ensure that it is not modified. Several questions to ask:

1. Is the data simple and pointer size or smaller? If so: use just `auto`, a copy is fine.
2. Are you doing a `move` or `initialization` where you want the data to be owned by the stack variable? just `auto` is fine.
3. Do you need a copy of the data to modify it? Or snapshot it at this moment in time? Just `auto`.
4. Do you need read-only access larger-than-pointer-size data? Use `const auto x&`?
5. Do you need to modify the data in place? Consider `auto x&` or use a pointer (smart pointer?) to modify the data.

Reply:
As for these things you mentioned I don't need the data in the auto to modify data in place or any other reason.
I just need the data for the moment and the data is pretty simple. I'd stay with `auto`.

Thanks for detailed explanation.